File: web/javascript/testCucumber/config/before.js
const { Before, setDefaultTimeout } = require("@cucumber/cucumber");
const playwright = require("playwright");
const selenium = require("selenium-webdriver");
const fs = require("fs").promises;
const path = require("path");

setDefaultTimeout(120000);

Before({ tags: "@playwright" }, async function (scenario) {
  // Clear old traces
  await manageDirectory();

  this.browser = await playwright.chromium.launch({
    headless: false,
    slowMo: 500,
  });
  this.context = await this.browser.newContext({
    defaultNavigationTimeout: 30000,
    defaultTimeout: 30000,
  });
  this.page = await this.context.newPage();
  await this.context.tracing.start({ screenshots: true, snapshots: true });

  console.log("Launching Playwright Chromium Browser");
});

// Selenium hook
Before({ tags: "@selenium" }, async function () {
  // Define your Selenium WebDriver settings here
  this.driver = new selenium.Builder()
    .forBrowser("chrome") // or 'firefox', 'edge', etc.
    .build();

  console.log("Launching Selenium WebDriver");
});

async function manageDirectory() {
  const directory = '.testArtifacts/trace';

  try {
      // Check if the directory exists
      try {
          await fs.access(directory);
          // If this succeeds, the directory exists - clear its contents
          const files = await fs.readdir(directory);
          for (const file of files) {
              await fs.unlink(path.join(directory, file));
          }
      } catch {
          // If an error is thrown, the directory does not exist - create it
          await fs.mkdir(directory, { recursive: true });
      }
  } catch (err) {
      console.error(`An error occurred: ${err}`);
  }
}

File: web/python/testCucumber/config/before.py
import os
from pathlib import Path
from playwright.sync_api import sync_playwright
from selenium import webdriver
from selenium.webdriver.chrome.service import Service as ChromeService
from webdriver_manager.chrome import ChromeDriverManager
from behave import before, use_step_matcher
from selenium.webdriver.firefox.service import Service as FirefoxService
from webdriver_manager.firefox import GeckoDriverManager

DEFAULT_TIMEOUT = 120000
use_step_matcher("re")

# Set default timeout for all scenarios
def before_all(context):
    print('hi')
    context.default_timeout = DEFAULT_TIMEOUT

# Playwright hook
@before(u'@playwright')
def before_playwright_scenario(context):
    # Clear old traces
    trace_dir = Path('./testArtifacts/trace')
    if trace_dir.exists():
        for file in trace_dir.iterdir():
            file.unlink()

    playwright = sync_playwright().start()
    context.browser = playwright.chromium.launch(headless=False)
    context.context = context.browser.new_context(
        default_navigation_timeout=30000,
        default_timeout=30000,
    )
    context.page = context.context.new_page()
    context.context.tracing.start(screenshots=True, snapshots=True)

    print('Launching Playwright Chromium Browser')

# Selenium hook
@before(u'@selenium')
def before_selenium_scenario(context):
    # Define your Selenium WebDriver settings here
    context.driver = webdriver.Chrome(service=ChromeService(ChromeDriverManager().install()))
    # Alternatively, for Firefox:
    # context.driver = webdriver.Firefox(service=FirefoxService(GeckoDriverManager().install()))

    print('Launching Selenium WebDriver')
